pulp_2to3_migration/app/models/base.py

```diff
@@ -330,7 +337,153 @@ def _parse_plugin_plan(self, repository_data):
+           the previous plan
+
+        If previous premigration failed, we should be careful not to override any in `new` state
+        with `up to date` ones, to let the `new` ones be processed (potentially for the second
+        time).
+
+        Args:
+            repo_id(str): pulp 2 repository id
+            repo_type(str): pulp 2 repo type
+            importer_repo_id(str): pulp 2 repository id of an importer
+        """
+        relation, created = cls.objects.get_or_create(
+            pulp2_resource_type=cls.IMPORTER,
+            pulp2_repo_type=repo_type,
+            pulp2_repo_id=repo_id,
+            pulp2_resource_repo_id=importer_repo_id,
+            defaults={'status': cls.NEW}
+        )
+
+        is_unchanged_relation = not created and relation.status == cls.OLD
+        if is_unchanged_relation:
+            relation.status = cls.UP_TO_DATE
+            relation.save()
+
+    @classmethod
+    def set_distributors(cls, repo_id, repo_type, distributor_repo_ids):
+        """
+        Sets proper status for the `repo_id`, `distributor_repo_id` pairs:
+         - `up to date` for the relations which stayed the same
+         - `new` for absolutely new ones or if a repository had different distributors according to
+           the previous plan
+
+        If previous premigration failed, we should be careful not to override any in `new` state
+        with `up to date` ones, to let the `new` ones be processed (potentially for the second
+        time).
+
+        Args:
+            repo_id(str): pulp 2 repository id
+            repo_type(str): pulp 2 repo type
+            distributor_repo_ids(list): pulp 2 repository ids of distributors
+        """
+        up_to_date_count = cls.objects.filter(
+            pulp2_repo_id=repo_id,
+            pulp2_resource_type=cls.DISTRIBUTOR,
+            pulp2_resource_repo_id__in=distributor_repo_ids,
+            status=cls.OLD
+        ).update(status=cls.UP_TO_DATE)
+
+        no_new_relations = up_to_date_count == len(distributor_repo_ids)
+        if no_new_relations:
+            return
+
+        # create new relations
+        for distributor_repo_id in distributor_repo_ids:
+            try:
+                cls.objects.create(
+                    pulp2_resource_type=cls.DISTRIBUTOR,
+                    pulp2_repo_type=repo_type,
+                    pulp2_repo_id=repo_id,
+                    pulp2_resource_repo_id=distributor_repo_id,
+                    status=cls.NEW,
+                )
+            except IntegrityError:
+                # ignore existing relations, they've been already updated as up to date ones
+                pass
+
+    @classmethod
+    def mark_changed_relations(cls):
+        """
+        Set is_migrated to False for any relations which changed and no longer up to date.
+        """
+        changed_relations_repo_ids = RepoSetup.objects.exclude(
+            status=cls.UP_TO_DATE
```

Comment:
@ipanova , this is a fix for the "moved" case, it used to mark only `new` ones and didn't look at `old` relations.